### PR TITLE
Merge materials

### PR DESCRIPTION
**Pull request by pkratten**
Merge materials on layer or object level.
Merge the default material.

This is to make the gltf more compact and organized.

Current behavior is that every object gets its own material if options.UseDisplayColorForUnsetMaterials is enabled; exceptions is sub-objects with different colors.
If options.ExportLayers is also enabled the layer color gets its own material for all (sub-)objects that are set to layer color.

The DefaultMaterial.RenderMaterial was giving different guids each call so i set it to 0.

### CODE REVIEW

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 1]
```diff
@@ -94,7 +98,7 @@ public Gltf ConvertToGltf()
 
                 int nodeIndex = dummy.Nodes.AddAndReturnIndex(node);
 
-                if(options.ExportLayers)
+                if (options.ExportLayers)
                 {
                     AddToLayer(RhinoDoc.ActiveDoc.Layers[exportData.Object.Attributes.LayerIndex], nodeIndex);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:103]: We need to remove these ActiveDoc calls since they aren't safe and could crash on Mac. I missed this one in the last PR. I've added a commit to properly deliver the RhinoDoc to the RhinoDocGltfConverter. Use that instead of these ActiveDoc calls.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 2]
```diff
@@ -202,6 +227,17 @@ int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
             return materialIndices;
         }
 
+        private int GetLayerMaterial(RhinoObject rhinoObject)
+        {
+            if (!layerMaterialIndices.TryGetValue(rhinoObject.Attributes.LayerIndex, out int layerMaterialIndex))
+            {
+                Color4f objectColor = GetObjectColor(rhinoObject);
+                layerMaterialIndex = CreateSolidColorMaterial(objectColor, RhinoDoc.ActiveDoc.Layers[rhinoObject.Attributes.LayerIndex].Name);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:235]: This one too

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 3]
```diff
@@ -218,11 +254,11 @@ int CreateSolidColorMaterial(Color4f color, string name)
 
         Color4f GetObjectColor(RhinoObject rhinoObject)
         {
-            if(rhinoObject.Attributes.ColorSource == ObjectColorSource.ColorFromLayer)
+            if (rhinoObject.Attributes.ColorSource == ObjectColorSource.ColorFromLayer)
             {
                 int layerIndex = rhinoObject.Attributes.LayerIndex;
 
-                return new Color4f(rhinoObject.Document.Layers[layerIndex].Color);
+                return new Color4f(RhinoDoc.ActiveDoc.Layers[layerIndex].Color);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:261]: And this one

**Review by jrz371 — CHANGES_REQUESTED** [review]

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 3 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: c879f6c0e12125a5f1b34176d1cfe80bdb12aa8a

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 2 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: c879f6c0e12125a5f1b34176d1cfe80bdb12aa8a

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 1 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: c879f6c0e12125a5f1b34176d1cfe80bdb12aa8a

**Comment by pkratten** [comment]
I don't know much about the mac side of things. Good to know, thanks!

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 4]
```diff
@@ -202,6 +227,17 @@ int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
             return materialIndices;
         }
 
+        private int GetLayerMaterial(RhinoObject rhinoObject)
+        {
+            if (!layerMaterialIndices.TryGetValue(rhinoObject.Attributes.LayerIndex, out int layerMaterialIndex))
+            {
+                Color4f objectColor = GetObjectColor(rhinoObject);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:234]: Should the layer material be relying on the object color? This won't return the layer color if the objects color source isn't from the layer.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 5]
```diff
@@ -165,8 +169,11 @@ public byte[] GetBinaryBuffer()
 
         int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
         {
+            RhinoObject[] subObjects = rhinoObject.GetSubObjects();
             int[] materialIndices = new int[materials.Length];
 
+            Dictionary<Color4f, int> colorIndices = new Dictionary<Color4f, int>();
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:175]: Shouldn't this Dictionary be a class member? If several objects share a color there will still be duplicate items if I'm reading this correctly. You should also use a different key than a Color4f. Perhaps an integer using the System.Drawing.Color.ToArgb(). Floating point precision may cause duplicate materials being stored. I'm also not sure the Color4f Hash function will work well in this situation.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 6]
```diff
@@ -176,18 +183,36 @@ int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
                     return null;
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:183]: This check should probably happen before GetMaterials gets called. At the very least it should be outside the for loop.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 7]
```diff
@@ -79,7 +83,7 @@ public Gltf ConvertToGltf()
 
             var sanitized = SanitizeRhinoObjects(objects);
 
-            foreach(ObjectExportData exportData in sanitized)
+            foreach (ObjectExportData exportData in sanitized)
             {
                 int[] materialIndices = GetMaterials(exportData.RenderMaterials, exportData.Object);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:88]: The export materials check should happen here, not inside the GetMaterials function.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 8]
```diff
@@ -176,18 +183,36 @@ int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
...
                 if (material == null && options.UseDisplayColorForUnsetMaterials)
                 {
-                    Color4f objectColor = GetObjectColor(rhinoObject);
-                    materialIndices[i] = CreateSolidColorMaterial(objectColor, GetObjectName(rhinoObject));
+                    if (options.ExportLayers)
+                    {
+                        if (subObjects[i].Attributes.ColorSource == ObjectColorSource.ColorFromLayer)
+                        {
+                            materialIndices[i] = GetLayerMaterial(rhinoObject);
+                            continue;
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:194]: In this context these continues here and on line 205 are confusing. Much of this loop needs to be made more coherent. It's hard to follow which conditions will set which material index.

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 9]
```diff
@@ -176,18 +183,36 @@ int[] GetMaterials(RenderMaterial[] materials, RhinoObject rhinoObject)
...
+                        colorIndex = CreateSolidColorMaterial(objectColor, GetObjectName(rhinoObject));
+                        colorIndices.Add(objectColor, colorIndex);
+                    }
+                    materialIndices[i] = colorIndex;
                     continue;
                 }
                 else if (material == null)
                 {
                     material = Rhino.DocObjects.Material.DefaultMaterial.RenderMaterial;
+                    materialId = new Guid();
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:210]: Should probably use the [Guid.Empty](https://docs.microsoft.com/en-us/dotnet/api/system.guid.empty?view=net-5.0) here for the zero guid.

**Review by jrz371 — CHANGES_REQUESTED** [review]
I think the biggest fix is making that for loop that gathers the material indices more coherent and understandable.

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 5 above)
Comment (pkratten) [reply glTF-BinExporter/RhinoDocGltfConverter.cs:178]: It depends. Should objects still have their own materials? Or should colors be shared for the while document?

My original intent was to merge sub-object materials. But we can extend it in the spirit of compact file size.

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 4 above)
Comment (pkratten) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: #d12d92a4ff07f463ad185d5563aa4b72eb971369

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 5 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: No, materials should be able to be shared. Especially simple materials with only the base color set. This was the original functionality when using the materialsMap to track them by Guid to index.

**Comment by pkratten** [comment]
Just letting you know that I found a problem that need a bit longer to fix.

**Comment by pkratten** [comment]
In my opinion this is good to go. Let me know if there is anything else!
